File: realDebridDeleter/realDebridDeleter.js
console.log("RD Plugin: Script Loaded (Fetch Mode)");

(function () {
    'use strict';

    const PLUGIN_ID = "realDebridDeleter";
    const TASK_NAME = "Delete From Cloud";
    const BUTTON_ID = "rd-delete-plugin-btn";

    // --- 1. NETWORK HELPER (Fetch Mode) ---
    async function runPluginTask(mode, payload) {
        // We construct the GraphQL query manually, which is guaranteed to work
        const mutation = `
            mutation RunTask($plugin_id: ID!, $task_name: String!, $args: [PluginArgInput!]) {
                runPluginTask(plugin_id: $plugin_id, task_name: $task_name, args: $args)
            }
        `;

        // Format arguments for Stash (everything must be a string)
        const args = [
            { key: "mode", value: { str: mode } },
            ...Object.keys(payload).map(key => {
                const val = payload[key];
                const valStr = typeof val === 'object' ? JSON.stringify(val) : String(val);
                return { key: key, value: { str: valStr } };
            })
        ];

        const variables = {
            plugin_id: PLUGIN_ID,
            task_name: TASK_NAME,
            args: args
        };

        try {
            console.log(`RD Plugin: Sending task '${mode}'...`, variables);

            const response = await fetch('/graphql', {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({ query: mutation, variables: variables })
            });

            const json = await response.json();

            if (json.errors) {
                console.error("RD Plugin GraphQL Error:", json.errors);
                return { error: json.errors[0].message };
            }

            const resultString = json.data.runPluginTask;

            // Parse the Sandwich: ###JSON_START### ... ###JSON_END###
            const regex = /###JSON_START###([\s\S]*?)###JSON_END###/;
            const match = resultString.match(regex);

            if (match && match[1]) {
                return JSON.parse(match[1]);
            } else {
                console.warn("RD Plugin: No JSON markers found in output:", resultString);
                // Fallback: try parsing raw string in case logging was disabled
                try { return JSON.parse(resultString); } catch(e) {}
                return { error: "Invalid output from backend. Check Stash Logs." };
            }
        } catch (e) {
            console.error("RD Plugin Network Error:", e);
            return { error: e.message };
        }
    }

    // --- 2. BUTTON LOGIC ---
    async function handleButtonClick(sceneId) {
        const btn = document.getElementById(BUTTON_ID);
        const originalHtml = btn.innerHTML;
        
        const resetBtn = () => {
            btn.style.opacity = "1";
            btn.innerHTML = originalHtml;
        };

        btn.style.opacity = "0.7";
        btn.innerHTML = '<span class="fa fa-spinner fa-spin"></span> Checking...';

        // STEP 1: CHECK
        const report = await runPluginTask("check", { scene_id: sceneId });

        if (report.error || !report.torrent_id) {
            alert("Scan Failed: " + (report.error || "Unknown Error"));
            resetBtn();
            return;
        }

        // STEP 2: CONFIRM
        let confirmMsg = "";
        let scenesToDelete = [sceneId];

        if (report.is_pack) {
            const others = report.related_scenes;
            confirmMsg = `⚠️ PACK DETECTED ⚠️\n\n`;
            confirmMsg += `Torrent: ${report.torrent_name}\n`;
            confirmMsg += `Contains ${report.video_file_count} video files.\n`;

            if (others.length > 0) {
                confirmMsg += `\nAlso deleting ${others.length} other Stash scenes:\n`;
                others.forEach(s => confirmMsg += `- ${s.title}\n`);
                scenesToDelete = scenesToDelete.concat(others.map(s => s.id));
            } else {
                confirmMsg += `\n(No other Stash scenes found in this folder)\n`;
            }
            confirmMsg += `\nDelete ALL from Cloud & Stash?`;
        } else {
            confirmMsg = `Delete single file?\n\n${report.torrent_name}`;
        }

        if (!confirm(confirmMsg)) {
            resetBtn();
            return;
        }

        // STEP 3: EXECUTE
        btn.innerHTML = '<span class="fa fa-spinner fa-spin"></span> Deleting...';
        const result = await runPluginTask("delete", {
            torrent_id: report.torrent_id,
            scene_ids: scenesToDelete
        });

        if (result.error) {
            alert("Delete Failed: " + result.error);
            btn.innerHTML = '<span class="fa fa-exclamation-triangle"></span> Error';
        } else {
            alert(`🗑️ Success! Deleted ${result.deleted_scenes} scenes.`);
            btn.remove(); 
        }
    }

    // --- 3. INJECTOR ---
    setInterval(() => {
        const path = window.location.pathname;
        const match = path.match(/\/scenes\/(\d+)$/);
        if (!match) return;
        const sceneId = match[1];

        const existingBtn = document.getElementById(BUTTON_ID);
        if (existingBtn) {
            if (existingBtn.dataset.sceneId !== sceneId) existingBtn.remove();
            else return;
        }

        const toolbar = document.querySelector('.scene-toolbar-group') || 
                        document.querySelector('.ml-auto') ||
                        document.querySelector('.SceneHeader-toolbar');

        if (toolbar) {
            const btn = document.createElement('button');
            btn.id = BUTTON_ID;
            btn.dataset.sceneId = sceneId;
            btn.className = "btn btn-danger";
            btn.title = "Delete from RealDebrid & Stash";
            btn.style.marginLeft = "10px";
            btn.onclick = () => handleButtonClick(sceneId);
            btn.innerHTML = '<span class="fa fa-trash"></span> Cloud Delete';
            toolbar.insertBefore(btn, toolbar.firstChild);
        }
    }, 500);

})();

File: realDebridDeleter/realDebridDeleter.py
import os
import sys
import json
import requests
import re

# CONFIGURATION
STASH_URL = "http://localhost:9999/graphql"
CONFIG_PATH = "/root/.stash/config.yml" 
PLUGIN_ID = "realDebridDeleter"
VIDEO_EXTENSIONS = {'.mp4', '.mkv', '.avi', '.wmv', '.mov', '.m4v', '.flv', '.webm', '.ts', '.iso'}

# --- OUTPUT HELPERS ---
def log(msg):
    """Prints to stderr (Stash Logs)."""
    sys.stderr.write(f"[RD-Plugin] {msg}\n")
    sys.stderr.flush()

def send_response(payload, req_id):
    """
    LOG SCRAPER ARCHITECTURE: 
    Write JSON to STDERR wrapped in unique tags.
    """
    if not req_id:
        log("CRITICAL: No Request ID provided.")
        sys.exit(1)

    json_str = json.dumps(payload)
    tag = f"###RD_RES_{req_id}###"
    
    # Print sandwich to logs
    sys.stderr.write(f"{tag}{json_str}{tag}\n")
    sys.stderr.flush()
    sys.exit(0)

def error_exit(msg, req_id=None):
    log(f"ERROR: {msg}")
    if req_id:
        send_response({"error": msg}, req_id)
    sys.exit(1)

# --- CONFIG & NETWORK ---
def get_api_key_from_config():
    if not os.path.exists(CONFIG_PATH): return None
    try:
        with open(CONFIG_PATH, 'r', encoding='utf-8') as f:
            for line in f:
                if line.strip().startswith('api_key:'):
                    return line.split(':', 1)[1].strip().strip('"').strip("'")
    except: pass
    return None

def get_stash_headers():
    headers = {"Content-Type": "application/json"}
    api_key = os.environ.get('STASH_API_KEY') or get_api_key_from_config()
    if api_key: headers["ApiKey"] = api_key
    return headers

def get_rd_api_key():
    query = "query Configuration { configuration { plugins } }"
    try:
        r = requests.post(STASH_URL, json={'query': query}, headers=get_stash_headers())
        if r.status_code != 200: return None
        data = r.json().get('data', {}).get('configuration', {}).get('plugins', {})
        return data.get(PLUGIN_ID, {}).get('rd_api_key')
    except: return None

# --- LOGIC ---
def get_scene_details(scene_id):
    query = """query FindScene($id: ID!) { findScene(id: $id) { id title files { path basename } } }"""
    try:
        r = requests.post(STASH_URL, json={'query': query, 'variables': {'id': scene_id}}, headers=get_stash_headers())
        return r.json().get('data', {}).get('findScene')
    except: return None

def find_sibling_scenes(folder_path):
    # Ensure no trailing slash for the query
    clean_search_path = folder_path.rstrip(os.sep)
    
    log(f"DEBUG: Searching Stash for siblings in: {clean_search_path}")

    query = """query FindScenesByPath($filter: SceneFilterType!) { 
        findScenes(scene_filter: $filter) { 
            scenes { id title files { path } } 
        } 
    }"""
    
    variables = {"filter": {"path": {"value": clean_search_path, "modifier": "INCLUDES"}}}
    
    try:
        r = requests.post(STASH_URL, json={'query': query, 'variables': variables}, headers=get_stash_headers())
        data = r.json()
        
        # Log raw response for debugging
        # log(f"DEBUG: Stash Raw Response: {json.dumps(data)}") 

        scenes = data.get('data', {}).get('findScenes', {}).get('scenes', [])
        log(f"DEBUG: Stash Query found {len(scenes)} potential matches.")

        siblings = []
        match_prefix = clean_search_path + os.sep
        
        for s in scenes:
            if not s['files']: continue
            path = s['files'][0]['path']
            
            # Log why we accept/reject each file
            # is_direct_child = os.path.dirname(path) == clean_search_path
            # is_nested_child = path.startswith(match_prefix)
            
            if os.path.dirname(path) == clean_search_path or path.startswith(match_prefix):
                 siblings.append({'id': s['id'], 'title': s['title']})
            else:
                 # log(f"DEBUG: Rejecting sibling {s['id']} path mismatch: {path}")
                 pass
                 
        log(f"DEBUG: Final sibling count: {len(siblings)}")
        return siblings
    except Exception as e:
        log(f"DEBUG: Error finding siblings: {e}")
        return []

def get_torrent_info(filename, full_path, token):
    headers = {'Authorization': f'Bearer {token}'}
    folder_name = os.path.basename(os.path.dirname(full_path))
    file_name = os.path.basename(full_path)
    
    search_term = folder_name
    if folder_name.lower() in ['__all__', 'cloud', 'data', 'realdebrid', 'zurg', 'movies', 'shows', 'default']:
        search_term = file_name

    log(f"Searching RD for: {search_term}")
    try:
        r = requests.get('https://api.real-debrid.com/rest/1.0/torrents?limit=100', headers=headers)
        if r.status_code != 200: return None
        torrents = r.json()
    except: return None

    target = None
    clean_search = search_term.lower()
    clean_file = os.path.splitext(file_name)[0].lower()

    for t in torrents:
        t_name = t['filename'].lower()
        if clean_search == t_name or clean_file in t_name:
            target = t
            break
            
    if not target: return None

    try:
        r2 = requests.get(f"https://api.real-debrid.com/rest/1.0/torrents/info/{target['id']}", headers=headers)
        if r2.status_code == 200: return r2.json()
    except: pass
    
    return target

# --- MODES ---
def execute_delete_mode(input_data, token, req_id):
    log("Entering Delete Mode")
    torrent_id = input_data.get('torrent_id')
    scene_ids_raw = input_data.get('scene_ids', '[]')
    
    try:
        scene_ids = json.loads(scene_ids_raw)
    except:
        scene_ids = [scene_ids_raw] if scene_ids_raw else []
    
    if not torrent_id or torrent_id == "undefined":
        error_exit("Missing valid torrent_id.", req_id)

    log(f"Deleting Torrent ID: {torrent_id}")
    headers = {'Authorization': f'Bearer {token}'}
    del_r = requests.delete(f"https://api.real-debrid.com/rest/1.0/torrents/delete/{torrent_id}", headers=headers)
    
    if del_r.status_code not in [204, 200]:
        error_exit(f"RD Delete Failed ({del_r.status_code})", req_id)

    query = """mutation SceneDestroy($id: ID!) { sceneDestroy(input: {id: $id, delete_file: false, delete_generated: true}) }"""
    deleted_count = 0
    for sid in scene_ids:
        log(f"Deleting Stash Scene: {sid}")
        r = requests.post(STASH_URL, json={'query': query, 'variables': {'id': sid}}, headers=get_stash_headers())
        if r.status_code == 200: deleted_count += 1

    send_response({"status": "success", "deleted_scenes": deleted_count}, req_id)

def execute_check_mode(scene_id, token, req_id):
    log(f"Entering Check Mode for Scene {scene_id}")
    scene = get_scene_details(scene_id)
    if not scene or not scene['files']:
        error_exit("Scene has no files", req_id)

    full_path = scene['files'][0]['path']
    folder_path = os.path.dirname(full_path)
    
    torrent = get_torrent_info(scene['files'][0]['basename'], full_path, token)
    if not torrent:
        error_exit("Torrent not found in RD history (No name match)", req_id)

    files = torrent.get('files', [])
    video_count = sum(1 for f in files if os.path.splitext(f['path'])[1].lower() in VIDEO_EXTENSIONS)
    
    log(f"Found Torrent: {torrent['filename']} (Videos: {video_count})")
    
    siblings = find_sibling_scenes(folder_path)
    related_scenes = [s for s in siblings if str(s['id']) != str(scene_id)]
    
    send_response({
        "status": "check_complete",
        "torrent_id": torrent['id'],
        "torrent_name": torrent['filename'],
        "video_file_count": video_count,
        "is_pack": video_count > 1,
        "related_scenes": related_scenes, 
        "folder_path": folder_path
    }, req_id)

# --- MAIN ---
if __name__ == "__main__":
    try:
        input_data = json.load(sys.stdin)
        args = input_data.get('args', {}) if 'args' in input_data else input_data
        
        mode = args.get('mode', 'check')
        req_id = args.get('req_id')
        token = get_rd_api_key()
        
        if not token: error_exit("RD API Key missing in Plugin Settings", req_id)

        if mode == 'delete':
            execute_delete_mode(args, token, req_id)
        else:
            sid = args.get('scene_id')
            if not sid: error_exit("No Scene ID provided", req_id)
            execute_check_mode(sid, token, req_id)

    except Exception as e:
        r_id = locals().get('req_id', None)
        error_exit(f"Script Crash: {str(e)}", r_id)